client/remoterepo.c

```diff
@@ -106,14 +106,94 @@ _handle_curl_cb(
     }
     return dwError;
 }
+uint32_t
+TDNFParseAndGetURLFromMetalink(
+    PTDNF pTdnf,
+    const char *pszRepo,
+    const char *pszFile
+    )
+{
+    metalink_error_t metalink_error;
+    metalink_t* metalink = NULL;
+    metalink_file_t **files;
+    metalink_parser_context_t *metalink_context = NULL;
+    metalink_resource_t** resources;
+    char buf[BUFSIZ] = {0};
+    int length = 0;
+    int fd = -1;
+    uint32_t dwError = 0;
+
+    if(!pTdnf ||
+       !pTdnf->pArgs ||
+       IsNullOrEmptyString(pszRepo) ||
+       IsNullOrEmptyString(pszFile))
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
 
+    metalink_context = metalink_parser_context_new();
+    if(metalink_context == NULL)
+    {
+        dwError = ERROR_TDNF_OUT_OF_MEMORY;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    fd = open(pszFile, O_RDONLY);
+    if(fd == -1)
+    {
+        dwError = errno;
+        BAIL_ON_TDNF_SYSTEM_ERROR(dwError);
+    }
+    while((length = read(fd, buf, sizeof(buf))) != 0)
```

Comment:
read call ret val validation is wrong, use while((length = read(fd, buf, sizeof(buf))) > 0)

From man page
On error, -1 is returned, and errno is set appropriately.  In this case, it is left unspecified whether the file position (if any) changes.

Reply:
Also, this is just a hunch, buf is of size BUFSIZ which 1024 I guess, if pszFile has more than BUFSIZ characters, read won't null terminate buf. So it will crash, you need to revisit this part of code.
Read sizeof(buf)-1 bytes every time and do a memset buf to 0 before every iteration is my suggestion.

Reply:
already made > 0 ,pls check latest change :)